### DISCUSSION TITLE
The V Machine Learning Roadmap and Ecosystem

### DISCUSSION THREAD

**Discussion by ulises-jeremias** [vlang/v#19371] [Ideas]
This issue defines a first iteration of the Machine Learning Roadmap. All the specific tasks will be linked to each corresponding repository and also contained on the ML Project [here](https://github.com/orgs/vlang/projects/2).

We can think a ML Ecosystem as a composition of five different categories: **Array Protocols**, **Foundation**, **Technique Specific**, **Domain-Specific** and **Application Specific**.

## Array Protocols

For this category we seek to have an alternative implementation to arrays in V. In V it is not possible to create a structure that uses the same operators as an array, so it is not possible to implement a structure that has the same protocol as an array.

However, we started working on [VTL](https://github.com/vlang/vtl), a library similar to [`numpy`](https://numpy.org/), [`Arraymancer`](https://github.com/mratsim/Arraymancer), among others that offers a `Ndarray` (in this case called Tensor).

## Foundation

On this category I identify two aspects: **Algorithms** and **Plotting tools**.
On this area we did a cool progress and we already have some libraries for this:

- V Scientific Library, [VSL](https://github.com/vlang/vsl)., is a library with a great variety of different modules. Although most modules offer pure-V definitions, VSL also provides modules that wrap known C libraries among other backends that allow high performance computing as an alternative.
- [vsl.plot](https://github.com/vlang/vsl/tree/master/plot) is the plot submodule of VSL that follows the structure of
[Plotly's graph_objects](https://plotly.com/python/graph-objects/) to plot data. Will provide more backends in the future and also a pure V implementation at some point.

### Thirdparty

- [erdetn.vplot](https://github.com/erdetn/vplot) is a wrapper for GNU Plot (gnuplot_i)

## Technique Specific

- [vsl.ml](https://github.com/vlang/vsl/tree/master/ml) has Kmeans, LinearRegression, KNN, Params Regression, NLP, and a Data wrapper with an observer pattern.
- . . .

### Thirdparty

- [hamnn](https://github.com/holder66/hamnn) is a machine learning library for classification using a nearest neighbor algorithm based on Hamming distances.
- [vervet](https://github.com/lucasrdrgs/vervet) is a Dataframe manipulation tool written in V

## Domain Specific

_TODO_

## Application Specific

_TODO_

**Comment by danieldaeschle** [comment]
A keras/tf implementation would also be nice!

**Comment by ulises-jeremias** [comment]
@danieldaeschle sure! an initial iteration of that will be covered when https://github.com/vlang/vtl/tree/main/nn gets released! it is still WIP

**Comment by ArtemkaKun** [comment]
@ulises-jeremias Hi, maybe we should close/update/move this issue?

**Comment by trufae** [comment]
What about adding bindings for the c api exposed from llama2.cpp?

**Comment by joe-conigliaro** [comment]
Imo a very important topic, since python is a clear leader in ML space due to the existing libraries (and ease of use). I think there is a great opportunity for another language to make a name for itself.

BTW have you seen https://github.com/tinygrad/tinygrad ?

**Reply by ulises-jeremias** [reply]
yes, was one of my inspirations on VTL 😊 just need more time/help on it

**Comment by ulises-jeremias** [comment]
will update this before the Hacktoberfest. Also will add issues to VSL and VTL to continue working adding functionalities for ML. Right now working on VTL.nn
